File: src/p198/house_robber.py
from copy import deepcopy
from typing import Dict, List


class SolutionRecursive:
    def helper(self, nums: List[int], visitable: List[bool]) -> int:
        if not any(visitable):
            return 0

        result = 0
        for i, num in enumerate(nums):
            if not visitable[i]:
                continue

            new_visitable = deepcopy(visitable)
            if i != 0:
                new_visitable[i - 1] = False
            new_visitable[i] = False

            if i != len(new_visitable) - 1:
                new_visitable[i + 1] = False

            result = max(result, num + self.helper(nums, new_visitable))

        return result

    def rob(self, nums: List[int]) -> int:
        return self.helper(nums, [True] * len(nums))


class Solution:
    def rob(self, nums: List[int]) -> int:
        subproblems = [0] * (len(nums) + 1)
        subproblems[0] = 0
        subproblems[1] = nums[0]

        for i in range(2, len(nums) + 1):
            pick = subproblems[i - 2] + nums[i - 1]
            skip = subproblems[i - 1]
            subproblems[i] = max(pick, skip)

        return subproblems[-1]


def test_p198() -> None:
    assert Solution().rob([1, 2, 3, 1]) == 4
    assert Solution().rob([2, 7, 9, 3, 1]) == 12
    assert (
        Solution().rob(
            [82, 217, 170, 215, 153, 55, 185, 55, 185, 232, 69, 131, 130, 102]
        )
        == 1082
    )


def main() -> None:
    test_p198()


if __name__ == "__main__":
    main()

File: src/p198/house_robber.cpp
#include <unordered_map>
#include <vector>

class Solution {
   public:
    int rob(std::vector<int>& nums) { return tabulationSpaceOptimized(nums); }

   private:
   // TLE
    int recursion(std::vector<int>& nums) { return recursionHelper(nums, 0); }
    int recursionHelper(const std::vector<int>& nums, const int index) {
        if (index >= nums.size()) {
            return 0;
        }

        int pick = nums[index] + recursionHelper(nums, index + 2);
        int skip = recursionHelper(nums, index + 1);

        return std::max(pick, skip);
    }

   // TLE
    int memoization(const std::vector<int>& nums) {
        std::unordered_map<int, int> cache;
        return memoizationHelper(nums, cache, 0);
    }

    int memoizationHelper(const std::vector<int>& nums,
                          std::unordered_map<int, int>& cache,
                          const int index) {
        if (cache.find(index) != cache.end()) {
            return cache[index];
        }

        int pick = nums[index] + recursionHelper(nums, index + 2);
        int skip = recursionHelper(nums, index + 1);

        int result = std::max(pick, skip);
        cache.insert({index, result});
        return result;
    }

   // O(N) time complexity, O(N) space complexity
    int tabulation(const std::vector<int>& nums) {
        std::vector<int> subProblems(nums.size() + 1, 0);
        subProblems[0] = 0;
        subProblems[1] = nums[0];
        for (int i = 1; i < nums.size(); i++) {
            int pick = subProblems[i-1] + nums[i];
            int skip = subProblems[i];
            subProblems[i+1] = std::max(pick, skip);
        }
        return subProblems.back();
    }

    // O(N) time complexity, O(1) space complexity
    int tabulationSpaceOptimized(const std::vector<int>& nums) {
        int twoAgo = 0, oneAgo = 0, result = 0;
        for (int i = 0; i < nums.size(); i++) {
            int pick = twoAgo + nums[i];
            int skip = oneAgo;
            result = std::max(pick, skip);

            twoAgo = oneAgo;
            oneAgo = result;
        }
        return result;
    }
};
